From b20039f4c8faf5517e68c23de97bde768fca938b Mon Sep 17 00:00:00 2001
From: "Lu.Jiang" <lu.jiang@windriver.com>
Date: Fri, 5 Sep 2014 13:47:29 +0800
Subject: [PATCH 0217/1089] powerpc/e6500: fully disable threads

[Original patch taken from QorIQ-SDK-V1.6-SOURCE-20140619-yocto.iso]

smt-enabled=off doesn't actually stop the secondary thread from being
enabled, as that's done early in a hard-coded way.  This is a hack to
prevent that from happening with minimal side effects.

With this patch, if CONFIG_PPC_DISABLE_THREADS=y, threads are
unconditionally disabled, without regard to the kernel command line.  If
you want to enable threads, you must build with
CONFIG_PPC_DISABLE_THREADS=n.

Signed-off-by: Scott Wood <scottwood@freescale.com>
Signed-off-by: Andy Fleming <afleming@freescale.com>
(cherry picked from commit 7d0a33c23e85e085dcf4e6cea47de714e8f34073)
Change-Id: I578cb4ae43fa6fab0f75b24ce763667dbeb72d69
Reviewed-on: http://git.am.freescale.net:8181/1706
Reviewed-by: Schmitt Richard-B43082 <B43082@freescale.com>
Reviewed-by: Fleming Andrew-AFLEMING <AFLEMING@freescale.com>
Tested-by: Fleming Andrew-AFLEMING <AFLEMING@freescale.com>
Signed-off-by: Lu.Jiang <lu.jiang@windriver.com>
---
 arch/powerpc/kernel/exceptions-64e.S   | 2 ++
 arch/powerpc/kernel/setup_64.c         | 4 ++++
 arch/powerpc/platforms/Kconfig.cputype | 6 ++++++
 3 files changed, 12 insertions(+)

diff --git a/arch/powerpc/kernel/exceptions-64e.S b/arch/powerpc/kernel/exceptions-64e.S
index 37efef8..c34ab3c 100644
--- a/arch/powerpc/kernel/exceptions-64e.S
+++ b/arch/powerpc/kernel/exceptions-64e.S
@@ -1393,6 +1393,7 @@ BEGIN_FTR_SECTION
 	andi.	r3,r3,2
 	bne	1b
 
+#ifndef CONFIG_PPC_DISABLE_THREADS
 	/* Configure the MSR per the default */
 	LOAD_REG_IMMEDIATE(r3, MSR_KERNEL);
 	MTTMR(TMRN_IMSR1, 3);
@@ -1407,6 +1408,7 @@ BEGIN_FTR_SECTION
 	/* Release the other thread. It will spin until kick_cpu is called */
 	li	r3, 2
 	mtspr	SPRN_TENS, r3
+#endif
 END_FTR_SECTION_IFSET(CPU_FTR_SMT)
 2:	blr
 
diff --git a/arch/powerpc/kernel/setup_64.c b/arch/powerpc/kernel/setup_64.c
index 00195c9..f2fada3 100644
--- a/arch/powerpc/kernel/setup_64.c
+++ b/arch/powerpc/kernel/setup_64.c
@@ -134,6 +134,9 @@ static char *smt_enabled_cmdline;
 /* Look for ibm,smt-enabled OF option */
 static void check_smt_enabled(void)
 {
+#ifdef CONFIG_PPC_DISABLE_THREADS
+	smt_enabled_at_boot = 0;
+#else
 	struct device_node *dn;
 	const char *smt_option;
 
@@ -171,6 +174,7 @@ static void check_smt_enabled(void)
 			of_node_put(dn);
 		}
 	}
+#endif
 }
 
 /* Look for smt-enabled= cmdline option */
diff --git a/arch/powerpc/platforms/Kconfig.cputype b/arch/powerpc/platforms/Kconfig.cputype
index 109d069..78fb09d 100644
--- a/arch/powerpc/platforms/Kconfig.cputype
+++ b/arch/powerpc/platforms/Kconfig.cputype
@@ -223,6 +223,12 @@ config FSL_ERRATUM_A_006184_PERIOD
 	  since when idle we will always have just executed from the main
 	  kernel mapping, so it should not be absent from the L1 I-MMU.
 
+config PPC_DISABLE_THREADS
+	bool "Avoid the use of hardware threads"
+	help
+	  Define this if running e6500 rev1 to avoid bugs
+	  relating to hardware threads.
+
 config PPC_FPU
 	bool
 	default y if PPC64
-- 
2.0.2

